build_sis_analyst_multi_turn_extended.py

```diff
@@ -0,0 +1,308 @@
+* error code: `{parsed_content['error_code']}`
+
+Message:
+```
+{parsed_content['message']}
+```
+        """
+        return parsed_content, error_msg
+
+
+def display_conversation():
+    """
+    Display the conversation history between the user and the assistant.
+    """
+    for idx, message in enumerate(st.session_state.messages):
+        role = message["role"]
+        content = message["content"]
+        with st.chat_message(role):
+            display_message(content, idx)
+
+
+def display_message(content: List[Dict[str, str]], message_index: int):
+    """
+    Display a single message content.
+
+    Args:
+        content (List[Dict[str, str]]): The message content.
+        message_index (int): The index of the message.
+    """
+    for item in content:
+        if item["type"] == "text":
+            st.markdown(item["text"])
+        elif item["type"] == "suggestions":
+            # Display suggestions as buttons
+            for suggestion_index, suggestion in enumerate(item["suggestions"]):
+                if st.button(
+                    suggestion, key=f"suggestion_{message_index}_{suggestion_index}"
+                ):
+                    st.session_state.active_suggestion = suggestion
+        elif item["type"] == "sql":
+            # Display the SQL query and results
+            display_sql_query(item["statement"], message_index)
+        else:
+            # Handle other content types if necessary
+            pass
+
+
+@st.cache_data(show_spinner=False)
+def get_query_exec_result(query: str) -> Tuple[Optional[pd.DataFrame], Optional[str]]:
+    """
+    Execute the SQL query and convert the results to a pandas DataFrame.
+
+    Args:
+        query (str): The SQL query.
+
+    Returns:
+        Tuple[Optional[pd.DataFrame], Optional[str]]: The query results and the error message.
+    """
+    global session
+    try:
+        df = session.sql(query).to_pandas()
+        return df, None
+    except SnowparkSQLException as e:
+        return None, str(e)
+
+
+def display_sql_query(sql: str, message_index: int):
+    """
+    Executes the SQL query and displays the results in form of data frame and charts.
+
+    Args:
+        sql (str): The SQL query.
+        message_index (int): The index of the message.
+    """
+
+    # Display the SQL query
+    with st.expander("SQL Query", expanded=False):
+        st.code(sql, language="sql")
+
+    # Display the results of the SQL query
+    with st.expander("Results", expanded=True):
+        with st.spinner("Running SQL..."):
+            df, err_msg = get_query_exec_result(sql)
+            if df is None:
+                st.error(f"Could not execute generated SQL query. Error: {err_msg}")
+                return
+
+            if df.empty:
+                st.write("Query returned no data")
+                return
+
+            # Show query results in two tabs
+            data_tab, chart_tab = st.tabs(["Data 📄", "Chart 📉"])
```

Comment:
nit: the line should only go up :)
```suggestion
            data_tab, chart_tab = st.tabs(["Data 📄", "Chart 📈"])
```